Show validation errors on step 2 of the onboarding flow when unable to continue

eason9487 opened the pull request:
### Changes proposed in this Pull Request:


This is a part of implementation for #634 and for 📌 [Display form validation errors](https://github.com/woocommerce/google-listings-and-ads/issues/1993#tasks-display-validation-errors) in #1993.

To show validation errors on step 2 of the onboarding flow when unable to continue, this PR:

#### Extract the shareable component and adjust the layouts of shipping time components

- Add a new component `ValidationErrors` for showing form validation error messages, and replace the errors rendering in `AssetGroupCard` with it.
- Move the Card layouts from `ShippingTimeSection` into `ShippingTimeSetup` and remove unneeded codes.

#### Fix and adjust validator `js/src/components/free-listings/configure-product-listings/checkErrors.js`

- Fix incorrect keys and rephrase a few messages.
- Update the validation of tax rate data to check if the store country code is `US`.

#### Add the mechanism of showing validation errors to the product listings form

- Request showing validation errors when clicking the submit of product listings form if any data is invalid.
- Add a function to render validation errors to the product listings form.
- Add the rendering of validation errors to:
   - `ChooseAudienceSection`
   - `FlatShippingRatesInputCards`
   - `EstimatedShippingRatesCard`
   - `OfferFreeShippingCard`
   - `MinimumOrderCard`
   - `ShippingTimeSetup`
   - `TaxRate`

### Screenshots:

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/4c977398-fb9b-4abc-9ef6-fc551cca362b

### Detailed test instructions:

Some free listings data are unable to clean via UI. It might need to run the following SQL to clean them from database.

```sql
DELETE FROM wp_options WHERE option_name = 'gla_target_audience';
DELETE FROM wp_options WHERE option_name = 'gla_merchant_center';
```

1. Go to step 2 of the onboarding flow.
2. Before clicking the "Continue" button, the UI should not show any validation errors even if there are invalid data.
3. Check if the "Continue" button is always enabled no matter whether the data are valid or not.
4. With some invalid data, click the "Continue" button to see if the validation errors are shown in the corresponding places.
5. Change data to view if the validation errors are shown or hidden accordingly.
6. Make all data valid and click the "Continue" button to check if it goes to step 3.

### Additional details:

💡 The implementation for step 3 will be done by subsequent PR(s)

### Changelog entry

eason9487 committed:
Add a new component `ValidationErrors` for showing form validation error messages

eason9487 committed:
Replace the errors rendering in `AssetGroupCard` with `ValidationErrors`

eason9487 committed:
Fix incorrect keys for the validator of product listings data and also rephrase a few messages

eason9487 committed:
Add the store country check to the validator of product listings data

eason9487 committed:
Move the Card layouts from `ShippingTimeSection` into `ShippingTimeSetup` and remove unneeded codes

eason9487 committed:
Request showing validation errors when clicking the submit of product listings form if any data is invalid

eason9487 committed:
Add a function to render validation errors to the product listings form

eason9487 committed:
Add the rendering of validation errors to `ChooseAudienceSection`

eason9487 committed:
Add the rendering of validation errors to shipping rates-related card components

eason9487 committed:
Add the rendering of validation errors to `ShippingTimeSetup`

eason9487 committed:
Add the rendering of validation errors to `TaxRate`

puntope commented on js/src/components/validation-errors/validation-errors.js:
```diff
@@ -0,0 +1,30 @@
+ * @param {Object} props React props
+ * @param {string|string[]} [props.messages] Validation error message(s).
+ */
+export default function ValidationErrors( { messages: rawMessages } ) {
+	let messages = rawMessages;
+
+	if ( ! rawMessages?.length ) {
+		return null;
+	}
+
+	if ( ! Array.isArray( rawMessages ) ) {
+		messages = [ rawMessages ];
+	}
+
+	return (
+		<ul className="gla-validation-errors">
+			{ messages.map( ( message ) => (
+				<li key={ message }>{ message }</li>
+			) ) }
+		</ul>
+	);
+}
```
This function was a bit hard for me to read, ideally, I would expect to only allow an array.

If that's not possible I think using another naming instead of `messages` for the local variable would make it easier to understand. At least for me, the fact that the local variable `messages` is the same as the argument `massages` an right after, we assigned it to `rawMessages` wich is the name override for `messages` made it hard to understand at first sight.

A potential way to do it might be:

```jsx
export default function ValidationErrors( { messages } ) {
        let messagesList = messages;  

        if ( ! messages?.length ) {
		return null;
	}
	
        if ( ! Array.isArray( messages ) ) {
		messagesList = [ messages ];
	}
	

	return (
		<ul className="gla-validation-errors">
			{ messagesList.map( ( message ) => (
				<li key={ message }>{ message }</li>
			) ) }
		</ul>
	);
}
```

eason9487 replied:
Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/2019/commits/10df591df6bdf0c0bc8ac68e399776c7cab7544a.

puntope commented:
> DELETE FROM wp_options WHERE option_name = 'gla_target_audience';
DELETE FROM wp_options WHERE option_name = 'gla_merchant_center';

I was able to perform the test without this. Might you explain why is this needed? Does the user (or us programmatically) handle this after the user upgrades with this feature?

puntope commented:
Is there any place where I can check the new validation error texts? Or there are no specific requirements?

puntope commented:
When I was testing the errors I discover some inconsistency with the Audience Selector and the rest of inputs.


https://github.com/woocommerce/google-listings-and-ads/assets/5908855/af27482e-c0f3-484b-b72e-726cae4c9425

As you can see in the video, the inputs are not showing validation errors until the validation error in the Audience selector is solved.

puntope reviewed (commented):
I check the requirements for this feature:

✅  Don’t disable the “Continue” or “Compete” buttons even if the form validation doesn’t pass.

✅  The form validation errors are not displayed until the user clicks the above button.

⚠️  I was unable to find a reference for the texts in the `errors.shipping_country_rates` and `errors. shipping_country_times`

⚠️  Some inconsistencies in the error renderization (see comments)

💅  A minor suggestion for the function `ValidationErrors()`

eason9487 committed:
Rename variables in the `ValidationErrors`

Address https://github.com/woocommerce/google-listings-and-ads/pull/2019#discussion_r1269425104

eason9487 committed:
Fix the unsynced value for `AppInputNumberControl` when clearing a value equivalent to 0.

eason9487 commented:
Hi @puntope, thanks for the review.

This PR is ready for the second round of code reviews. Could you help with it?

> > DELETE FROM wp_options WHERE option_name = 'gla_target_audience';
> > DELETE FROM wp_options WHERE option_name = 'gla_merchant_center';
> 
> I was able to perform the test without this. Might you explain why is this needed? Does the user (or us programmatically) handle this after the user upgrades with this feature?

The selected **Location** option value ("Selected countries only" or "All countries") is stored in the `gla_target_audience`. The selected **Shipping rates** and **Tax rate (required for U.S. only)** option values are stored in the `gla_merchant_center`. These values are unable to be cleared via the UI. Deleting them from DB is a way to test step 2 as in the case of the first time entering the onboarding flow.

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/fc82a595-b21c-490a-b56d-4fab94694ec3)

P.S. **Location** and **Shipping rates** will be filled in automatically when empty, so this PR didn't add validation errors for them but it's still a bit better to double-confirm that the automatic file in processing works.

> Is there any place where I can check the new validation error texts? Or there are no specific requirements?

No. Except for PMax Assets and phone verification, no previous development or design documents have specified the texts for validation errors.

The main reason for adjusting some texts was to avoid the following situation, which happens to be a line break at the last word.

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/776a4ab6-ef78-4ada-b4d8-19f43bde8157)

> When I was testing the errors I discover some inconsistency with the Audience Selector and the rest of inputs. [...] As you can see in the video, the inputs are not showing validation errors until the validation error in the Audience selector is solved.

This is expected, as each field focuses on different validation errors. When no audience country is selected, the relevant validation errors are already displayed in the most relevant position. It would be a bit redundant to display the same error on both Shipping rates and Shipping times.

However, I noticed an issue from your video: after clearing a value equal to `0` in Shipping rates or Shipping times and moving out the focus, the `AppInputNumberControl` should be backfilled with 0 again but it didn't.

I tried a couple of different approaches and finally realized that there was only one way to solve the issue first with a relatively small amount of changes, while the rest of the approaches involved too many changes. Please view the fix in https://github.com/woocommerce/google-listings-and-ads/pull/2019/commits/a8f9bf90a5b864956a8fe784582907b8aa850dd1.

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/4d5237d3-f759-417b-a499-00b6af1de3c9

puntope commented on js/src/components/app-input-number-control/index.js:
```diff
@@ -36,6 +41,17 @@ const AppInputNumberControl = ( props ) => {
 
+	// The `InputControl` in `@wordpress/components` doesn't take into account the context
+	// in which the `value` of a controlled component is updated via `onBlur`, resulting in
+	// the incoming `value` prop may not syncing to the <input> element's display value.
+	// Therefore, a synchronization needle is used to make the value synchronized via the
+	// onChange callback.
```
Thanks for the changes. They look good. The only thing is that "sync Needle" naming, seems to me a kinda personal concept name? . I was checking around the internet and I didn't find anything regarding "synchronisation needle". So I would suggest using something more descriptive like `refreshInput`

eason9487 replied:
The naming was inspired by the reset needle.

![resetrouter](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/a301cb0e-fc4f-47e4-8ece-efca081fb117)

Changed to adopt [an official approach and naming](https://legacy.reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate) in https://github.com/woocommerce/google-listings-and-ads/pull/2019/commits/f763449a5b47efffe9b3ddd3a9a12dc711563a56.

puntope replied:
Much better, thanks for this and for the reference.

puntope reviewed (approved):
✅  LGTM

Thanks for the adjustments, explanation and extra Unit tests.

I left one last 💅  comment in regards to that "syncNeedle" name.

eason9487 committed:
Change to adopt the force update approach as proposed in the React official document for `AppInputNumberControl`

Address https://github.com/woocommerce/google-listings-and-ads/pull/2019#discussion_r1272059836
